pep-0624.rst

```diff
@@ -0,0 +1,233 @@
+``PyUnicode_EncodeUnicodeEscape()``       ``PyUnicode_AsUnicodeEscapeString()``
+``PyUnicode_EncodeRawUnicodeEscape()``    ``PyUnicode_AsRawUnicodeEscapeString()``
+``PyUnicode_EncodeCharmap()``             ``PyUnicode_AsCharmapString()`` \(1)
+``PyUnicode_TranslateCharmap()``          ``PyUnicode_Translate()``
+``PyUnicode_EncodeDecimal()``              \(3)
+``PyUnicode_TransformDecimalToASCII()``    \(3)
+========================================= ====================================================================
+
+Notes:
+
+(1)
+   ``const char *errors`` is missing.
+
+(2)
+   ``const char *errors, int byteorder`` are missing.
+
+(3)
+   There is no direct replacement. But CPython uses
+   ``_PyUnicode_TransformDecimalAndSpaceToASCII``
+   for converting from Unicode to numbers instead.
+
+
+Plan
+====
+
+Python 3.9
+----------
+
+Add ``Py_DEPRECATED(3.3)`` to these APIs.
+
+* ``PyUnicode_EncodeDecimal()``
+* ``PyUnicode_TransformDecimalToASCII()``.
+
+This change is committed already.
+https://github.com/python/cpython/commit/9c3840870814493fed62e140cfa43c2883e12181
+
+All other APIs have been marked ``Py_DEPRECATED(3.3)`` already.
+
+Python 3.10
+-----------
+
+Remove these APIs.
+
+* ``PyUnicode_Encode()``
+* ``PyUnicode_EncodeASCII()``
+* ``PyUnicode_EncodeLatin1()``
+* ``PyUnicode_EncodeUTF7()``
+* ``PyUnicode_EncodeUTF8()``
+* ``PyUnicode_EncodeUTF16()``
+* ``PyUnicode_EncodeUTF32()``
+* ``PyUnicode_EncodeUnicodeEscape()``
+* ``PyUnicode_EncodeRawUnicodeEscape()``
+* ``PyUnicode_EncodeCharmap()``
+* ``PyUnicode_TranslateCharmap()``
+
+
+Python 3.11
+-----------
+
+Remove these APIs.
+
+* ``PyUnicode_EncodeDecimal()``
+* ``PyUnicode_TransformDecimalToASCII()``
+
+
+Alternative ideas
+=================
+
+Make some private APIs public
+------------------------------
+
+``PyUnicode_EncodeUTF7()`` doesn't have public alternative APIs.
+
+Some APIs have alternative public APIs. But they are missing
+``const char *errors`` or ``int byteorder`` parameters.
+
+We can rename some private APIs and make them public to cover missing
+APIs and parameters.
+
+=============================== ================================
+ Rename to                       Rename from
+=============================== ================================
+``PyUnicode_EncodeASCII()``      ``_PyUnicode_AsASCIIString()``
+``PyUnicode_EncodeLatin1()``     ``_PyUnicode_AsLatin1String()``
+``PyUnicode_EncodeUTF7()``       ``_PyUnicode_EncodeUTF7()``
+``PyUnicode_EncodeUTF8()``       ``_PyUnicode_AsUTF8String()``
+``PyUnicode_EncodeUTF16()``      ``_PyUnicode_EncodeUTF16()``
+``PyUnicode_EncodeUTF32()``      ``_PyUnicode_EncodeUTF32()``
+=============================== ================================
+
+Pros:
+
+* We can have more consistent API set.
+
+Cons:
+
+* We have more public stable APIs to maintain.
+* Existing public APIs are enough for most use cases, and
+  ``PyUnicode_AsEncodedString()`` can be used in other cases.
```

Comment:
I would add that the main advantage of specialized functions is performance, but so far, there was no explicit user request nor benchmark showing a significant speedup.

Reply:
Deprecated APIs are much slower because they creates temporary Unicode object by `PyUnicode_FromWideChar`.

In case of utf-16-le, utf-16-be, utf-32-le, utf-32-be, `PyUnicode_AsEncodedString()` need to search codec from registry. We may be able to add fast paths for them.

Reply:
An extension module cannot look for the encode and decode functions once, and then use it? The io.TextIOWrapper does that for example.